### PR TITLE
[Fix] streaming: do not reset timeout for each chunk emit

### PR DESCRIPTION
**Pull request by MehakBindra**
Issue : https://github.com/microsoft/teams.ts/issues/374

Main change::
- In emit, instead of cancelling and scheduling a new timeout, 
      - flush immediately if no timeout pending, 
      - else push to queue and wait.
- Added tests for streaming

https://github.com/user-attachments/assets/d8fc0a16-21f9-4ddc-a681-5db5e200e192

Devtools: Streaming always starts before the full response is received (check with logs, see attached video).
Teams: Even though the first chunk is emitted immediately, (sometimes) by the time the stream starts on Teams, the full response is ready. In the second msg, we can see the stream starts a little bit before the full response is printed.

### CODE REVIEW

**Review by corinagum — APPROVED** [review]

File: packages/apps/src/plugins/http/stream.ts [hunk 1]
```diff
@@ -76,15 +104,23 @@ export class HttpStream implements IStreamer {
...
     }
 
-    while (!this.id || this.queue.length) {
-      await new Promise((resolve) => setTimeout(resolve, 200));
+    // Wait until all queued activities are flushed
+    const start = Date.now();
+
+    while (this.queue.length || !this.id) {
+      if (Date.now() - start > this._totalTimeout) {
+        this._logger.warn('Timeout while waiting for id and queue to flush');
```
Comment (heyitsaamir) [inline packages/apps/src/plugins/http/stream.ts:112]: shouldn't you break out when this happens?

Comment (heyitsaamir) [reply packages/apps/src/plugins/http/stream.ts:112]: let's also make sure you test this behavior

File: packages/apps/src/utils/promises/retry.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ export async function retry<T = any>(factory: () => Promise<T>, options?: RetryO
   try {
     return await factory();
   } catch (err) {
-    if (max > 0) {
+    if (max > 1) {
```
Comment (heyitsaamir) [inline packages/apps/src/utils/promises/retry.ts:30]: why max > 1?

File: packages/apps/src/plugins/http/stream.spec.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
+import { HttpStream } from './stream';
```
Comment (heyitsaamir) [inline packages/apps/src/plugins/http/stream.spec.ts:1]: So these are pretty simple test cases. If you want to test anything that tests the edge cases, you'll need to mimic ticks.

Basically:
Emit
Advance Time by x
Emit
Advance Time by x
Assert if activity was sent as expected
Assert if activity contains what is expected (chunk size?)
Emit
Advance Time by y
Assert if activity is not sent

etc.

This is how time-related activities are tested. Checkout this gist that displays a lot of these - https://gist.github.com/apieceofbart/e6dea8d884d29cf88cdb54ef14ddbcc4

Anyway, as rido said, don't sink too much time in this, but something to consider.

File: packages/apps/src/utils/promises/retry.ts (hunk 2 above)
Comment (MehakBindra) [reply packages/apps/src/utils/promises/retry.ts:30]: max decreases by 1 for each attempt
max = no of attempts 
while max > 1, try again . (then no of tries = max )

File: packages/apps/src/plugins/http/stream.spec.ts (hunk 3 above)
Comment (MehakBindra) [reply packages/apps/src/plugins/http/stream.spec.ts:1]: updated some tests to use advanceTimersByTimeAsync instead of runAllTimersAsync to test time constraints better

**Review by singhk97 — APPROVED** [review]
